### DISCUSSION TITLE
Lambda Snapstart with Serverless framework

### DISCUSSION THREAD

**Discussion by mamakancha** [serverless/serverless#11570] [General]
So AWS announced [Lambda Snapstart][1] very recently, I tried to give it a go since my application has cold start time ~4s. 

I was able to do this by adding the following under `resources`:
```
- extensions:
      NodeLambdaFunction:
        Properties:
          SnapStart:
            ApplyOn: PublishedVersions
```

Now, when I actually go to the said lambda, this is what I see :

[![enter image description here][2]][2]

So far so good! 
But, the issue is that when I check my Cloudwatch Logs, there's no trace of `Restore Time` instead the good old `Init Duration` for cold starts which means Snapstart isn't working properly.

I dug deeper, so Snapstart only works for versioned ARNs. But the thing is Serverless already [claims][3] that :

> By default, the framework creates function versions for every deploy. 

And on checking the logs, I see that the logStreams have the prefix : `2022/11/30/[$LATEST]`. 

When I check the Versions tab in console, I see version number **240**. So I would expect that 240 is the latest version of this lambda function and this is the function version being invoked everytime.

However, clicking on the version number open a lambda function with **240** attached to its ARN and testing that function with Snapstart works perfectly fine.

So I am confused if the **LATEST** version and version number **240** ( in my case ), are these different? 

 - If no, then why isn't Snapstart automatically activated for LATEST?
 - If yes, how do I make sure they are same?

  [1]: https://aws.amazon.com/about-aws/whats-new/2022/11/aws-lambda-snapstart-java-functions/
  [2]: https://i.stack.imgur.com/KQ0tX.png
  [3]: https://www.serverless.com/framework/docs/providers/aws/guide/functions

**Comment by mamakancha** [comment]
Found a [source](https://lumigo.io/blog/aws-lambda-cold-starts-are-about-to-get-faster/) where it's clearly mentioned that :
 >SnapStart only works with Lambda versions, not the $LATEST alias. So when invoking the function, you need to use the full ARN of the SnapStart-enabled version.

Same thing was mentioned in official [AWS Documentation](https://docs.aws.amazon.com/lambda/latest/dg/snapstart-activate.html)
>Active – Snapshot creation is complete and you can invoke the function. To use SnapStart, you must invoke the published function version, not the unpublished version ($LATEST).

So, obviously I need a way for my lambda function to invoke a given version of lambda function ( 240  in my case ). Any ideas on how this could be done? Right now $LATEST gets invoked by deafault. Is there a way I can override this in my `.yml` file?

**Comment by ShubamVirdi** [comment]
Hey i am also facing the same issue. Did you guys were able to configure this via serverless.yml file

**Comment by wietlol** [comment]
With serverless framework, you can extend the AWS::Lambda::Function resources by using the resources cloudformation part of the serverless file.
```
functions:
    myFunc:
        handler: '...'

resources:
    Resources:
        # note that this is the PascalCase version of the function name with the postfix "LambdaFunction"
        # check your cloudformation stack resources if you are unsure what the name of your function resource is
        MyFuncLambdaFunction:
            Type: AWS::Lambda::Function
            Properties:
                SnapStart:
                    ApplyOn: PublishedVersions
```

This will enable SnapStart for your lamba function on published versions, which at the time of writing is the only place where this is supported.
By default, serverless will create new published versions on every deploy.
So unless that is disabled ( https://www.serverless.com/framework/docs/providers/aws/guide/functions#versioning-deployed-functions ) you should already see SnapStart enabled published versions after you deploy.

The issue is that you probably do not use published versions yet, which means that everything is using the unqualified $LATEST containers.

Published versions work a bit different than the unqualified ($LATEST) version and when calling the function, for example with the SDK (LambdaClient), you have to explicitly state which version you want to use.
This would mean that your function arn changes on every deploy, so this would be rather annoying.
However, you can create your own alias which can be moved to a different published version.
The easiest way would be to use https://github.com/digio/serverless-simple-alias and adding the following into your serverless.yml file:
```
custom:
    simpleAlias:
        activeAliasName: 'latest'
        aliases: []
```
This will create an alias 'latest' (note that it is unrelated to the default '$LATEST' alias) and it will keep this alias on the last published version.
Now you can invoke your functions using the 'latest' alias and it will use the SnapStart enabled containers.
This alias will always remain the same, so you dont have to update all invoking applications when you publish a new version.
```
// Kotlin code, but it works the same in Java
val response = lambdaClient.invoke {
	it.functionName("myFunc:latest") // either append your alias to your functionName or use the qualifier to invoke it
}
```

The issue is now that this will only use SnapStart containers if you use the AWS SDK to invoke your function.
When your function is a handler from SQS, SNS, API Gateway, etc, the published version will not be invoked.
You will have to tell all those services to invoke the latest aliased version.
For API Gateway, this should be easy as the plugin has built-in support to change the API Gateway resources to use your alias.
For the others, you have to put some stuff in the serverless.yml resources part in order to specify that the handlers should be the latest.
For example:
```
MySnsSubscription:
    Type: AWS::SNS::Subscription
    DependsOn:
        - MyFuncLambdaFunctionAlias
    Properties:
        Endpoint: arn:aws:lambda:${self:provider.region}:${self:custom.accountNumber}:function:myFunc:latest
        Protocol: lambda
        TopicArn: arn:aws:sns:${self:provider.region}:${self:custom.accountNumber}:MyTopic
```
